### PR TITLE
Simplifies member attributes

### LINKED ISSUE #14907: Simplify Member Attributes
Currently, member attributes support builtin types. It causes lots of duplication in the API and internal code with really minimal gain. See the API of Member 
```
...
    String getStringAttribute(String key);
    void setStringAttribute(String key, String value);
    Boolean getBooleanAttribute(String key);
    void setBooleanAttribute(String key, boolean value);
    Byte getByteAttribute(String key);
    void setByteAttribute(String key, byte value);
    Short getShortAttribute(String key);
    void setShortAttribute(String key, short value);
    Integer getIntAttribute(String key)
    void setIntAttribute(String key, int value);
    Long getLongAttribute(String key);
    void setLongAttribute(String key, long value)
    Float getFloatAttribute(String key);
    void setFloatAttribute(String key, float value);
    Double getDoubleAttribute(String key);
    void setDoubleAttribute(String key, double value);
...
```

I suggest providing only String to String attributes.

### PR DESCRIPTION
**Pull request by mustafaiman**
Removes primitives support for [set|get]Attribute operations. All member attributes including DiscoveryStrategy#discoverLocalMetadata and spring supports only string attributes now.

Fixes #14907

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/operations/MemberAttributeChangedOp.java [hunk 1]
```diff
@@ -30,12 +30,12 @@
 
     private MemberAttributeOperationType operationType;
     private String key;
-    private Object value;
+    private String value;
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/operations/MemberAttributeChangedOp.java]: You can use `writeUTF/readUTF` when serializing.

File: hazelcast/src/main/java/com/hazelcast/spi/discovery/DiscoveryStrategy.java [hunk 2]
```diff
@@ -91,5 +91,5 @@
      * @return a map of discovered metadata as provided by the runtime environment
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/spi/discovery/DiscoveryStrategy.java]: Not related to this PR but the javadoc seems wrong:
```
* Information discovered from this method are shaded into the {@link com.hazelcast.core.Member}s
     * attributes. Existing attributes will not be overridden, that way local attribute configuration
     * overrides provided metadata.
```
I see the metadata overriding the attributes. Can you confirm and change the javadoc while you're at it?

File: hazelcast/src/main/java/com/hazelcast/spi/discovery/DiscoveryNode.java [hunk 3]
```diff
@@ -71,5 +71,5 @@
      *
      * @return assigned properties of that node
      */
-    public abstract Map<String, Object> getProperties();
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/spi/discovery/DiscoveryNode.java]: Do you need to change the type of these as well? Are they connected to member attributes in any way?

File: hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/operations/MemberAttributeChangedOp.java (hunk 1 above)
Comment (mustafaiman) [reply hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/operations/MemberAttributeChangedOp.java]: good catch! thanks

File: hazelcast/src/main/java/com/hazelcast/spi/discovery/DiscoveryNode.java (hunk 3 above)
Comment (mustafaiman) [reply hazelcast/src/main/java/com/hazelcast/spi/discovery/DiscoveryNode.java]: No, they could stay unchanged. It is changed as a result of chain refactoring,

File: hazelcast/src/main/java/com/hazelcast/spi/discovery/DiscoveryStrategy.java (hunk 2 above)
Comment (mustafaiman) [reply hazelcast/src/main/java/com/hazelcast/spi/discovery/DiscoveryStrategy.java]: Yes, looks like they are overridden. I updated the Javadoc. Also, added a test in `DiscoverySpiTest.java` to verify member attributes are overridden by discovery attributes.

File: hazelcast-client/src/test/java/com/hazelcast/client/spi/impl/discovery/ClientDiscoverySpiTest.java [hunk 4]
```diff
@@ -429,7 +429,7 @@ public PartitionGroupStrategy getPartitionGroupStrategy() {
         }
 
         @Override
-        public Map<String, Object> discoverLocalMetadata() {
+        public Map<String, String> discoverLocalMetadata() {
```
Comment (sancar) [inline hazelcast-client/src/test/java/com/hazelcast/client/spi/impl/discovery/ClientDiscoverySpiTest.java]: This also seems unrelated.

**Comment by mesutcelik** [comment]
This PR breaks all discovery plugins when they move to hazelcast 4.0. We need to mention that in our migration plan. Following plugins can not compile against hazelcast 4.0-SNAPSHOT anymore. cc: @Serdaro @dbrimley @mmedenjak 

- hazelcast-kubernetes
- hazelcast-aws
- hazelcast-azure
- hazelcast-gcp
- hazelcast-eureka

I know there are a bunch of other community plugins too.

Those plugins have to make releases with hazelcast 4.0 Beta release...

**Comment by mmedenjak** [comment]
@mesutcelik is this because of the change in [DiscoveryStrategy](https://github.com/hazelcast/hazelcast/blob/6ded404db9a829f4ed9787288941153638ffaf2f/hazelcast/src/main/java/com/hazelcast/spi/discovery/DiscoveryStrategy.java#L94)?

**Comment by mesutcelik** [comment]
`SimpleDiscoveryNode` is the one apparently...

```
[ [1;31mERROR [m] <http://jenkins.hazelcast.com/job/Eureka-master/ws/src/main/java/com/hazelcast/eureka/one/EurekaOneDiscoveryStrategy.java>:[301,27] no suitable constructor found for SimpleDiscoveryNode(com.hazelcast.nio.Address,java.util.Map<java.lang.String,java.lang.Object>)
    constructor com.hazelcast.spi.discovery.SimpleDiscoveryNode.SimpleDiscoveryNode(com.hazelcast.nio.Address,java.util.Map<java.lang.String,java.lang.String>) is not applicable
      (argument mismatch; java.util.Map<java.lang.String,java.lang.Object> cannot be converted to java.util.Map<java.lang.String,java.lang.String>)
    constructor com.hazelcast.spi.discovery.SimpleDiscoveryNode.SimpleDiscoveryNode(com.hazelcast.nio.Address,com.hazelcast.nio.Address) is not applicable
      (argument mismatch; java.util.Map<java.lang.String,java.lang.Object> cannot be converted to com.hazelcast.nio.Address)
[ [1;31mERROR [m] <http://jenkins.hazelcast.com/job/Eureka-master/ws/src/main/java/com/hazelcast/eureka/one/EurekaOneDiscoveryStrategy.java>:[313,19] no suitable constructor found for SimpleDiscoveryNode(com.hazelcast.nio.Address,java.util.Map<java.lang.String,java.lang.Object>)
    constructor com.hazelcast.spi.discovery.SimpleDiscoveryNode.SimpleDiscoveryNode(com.hazelcast.nio.Address,java.util.Map<java.lang.String,java.lang.String>) is not applicable
      (argument mismatch; java.util.Map<java.lang.String,java.lang.Object> cannot be converted to java.util.Map<java.lang.String,java.lang.String>)
    constructor com.hazelcast.spi.discovery.SimpleDiscoveryNode.SimpleDiscoveryNode(com.hazelcast.nio.Address,com.hazelcast.nio.Address) is not applicable
      (argument mismatch; java.util.Map<java.lang.String,java.lang.Object> cannot be converted to com.hazelcast.nio.Address)
```

**Comment by mustafaiman** [comment]
@mesutcelik @mmedenjak properties of DiscoveryNode still supports Object properties. It is the only constructor of SimpleDiscoveryNode that was changed accidentally. Should be fixed by changing the constructor accordingly.

**Comment by mustafaiman** [comment]
Still DiscoveryStrategy#discoverLocalMetadata is changed and that requires a change in plugins I guess.

**Comment by mesutcelik** [comment]
Yes. That is true. However, it is only used by plugins which support ZONE_AWARE feature like hazelcast-kubernetes....

```
[ [1;31mERROR [m] <http://jenkins.hazelcast.com/job/Kubernetes-master/ws/src/main/java/com/hazelcast/kubernetes/HazelcastKubernetesDiscoveryStrategy.java>:[67,32] discoverLocalMetadata() in com.hazelcast.kubernetes.HazelcastKubernetesDiscoveryStrategy cannot implement discoverLocalMetadata() in com.hazelcast.spi.discovery.DiscoveryStrategy
  return type java.util.Map<java.lang.String,java.lang.Object> is not compatible with java.util.Map<java.lang.String,java.lang.String>
[ [1;31mERROR [m] <http://jenkins.hazelcast.com/job/Kubernetes-master/ws/src/main/java/com/hazelcast/kubernetes/HazelcastKubernetesDiscoveryStrategy.java>:[66,5] method does not override or implement a method from a supertype
[ [1;31mERROR [m] <http://jenkins.hazelcast.com/job/Kubernetes-master/ws/src/main/java/com/hazelcast/kubernetes/KubernetesApiEndpointResolver.java>:[76,113] incompatible types: java.util.Map<java.lang.String,java.lang.Object> cannot be converted to java.util.Map<java.lang.String,java.lang.String>
```
